internal/controller/servicetemplate_controller.go

```diff
@@ -0,0 +1,580 @@
+	case "Secret":
+		secret := &corev1.Secret{}
+		err = r.Get(ctx, key, secret)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred Secret %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromLocalObject(secret)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from Secret %s: %w", key, err)
+		}
+	case "ConfigMap":
+		configMap := &corev1.ConfigMap{}
+		err = r.Get(ctx, key, configMap)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred ConfigMap %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromLocalObject(configMap)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from ConfigMap %s: %w", key, err)
+		}
+	case "GitRepository":
+		gitRepository := &sourcev1.GitRepository{}
+		err = r.Get(ctx, key, gitRepository)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred GitRepository %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromFluxObject(gitRepository)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from GitRepository %s: %w", key, err)
+		}
+		conditions := make([]metav1.Condition, len(gitRepository.Status.Conditions))
+		copy(conditions, gitRepository.Status.Conditions)
+		status.SourceStatus.Conditions = conditions
+	case "Bucket":
+		bucket := &sourcev1.Bucket{}
+		err = r.Get(ctx, key, bucket)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred Bucket %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromFluxObject(bucket)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from Bucket %s: %w", key, err)
+		}
+		conditions := make([]metav1.Condition, len(bucket.Status.Conditions))
+		copy(conditions, bucket.Status.Conditions)
+		status.SourceStatus.Conditions = conditions
+	case "OCIRepository":
+		ociRepository := &sourcev1beta2.OCIRepository{}
+		err = r.Get(ctx, key, ociRepository)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred OCIRepository %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromFluxObject(ociRepository)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from OCIRepository %s: %w", key, err)
+		}
+		conditions := make([]metav1.Condition, len(ociRepository.Status.Conditions))
+		copy(conditions, ociRepository.Status.Conditions)
+		status.SourceStatus.Conditions = conditions
+	default:
+		err = fmt.Errorf("unsupported source kind %s", ref.Kind)
+	}
+
+	if err != nil {
+		status.TemplateValidationStatus.Valid = false
+		status.TemplateValidationStatus.ValidationError = err.Error()
+	} else {
+		status.TemplateValidationStatus.Valid = true
+	}
+	template.Status = status
+	return err
+}
+
+func (r *ServiceTemplateReconciler) reconcileRemoteSource(ctx context.Context, template *kcm.ServiceTemplate) error {
+	ref := template.RemoteSourceSpec()
+	if ref == nil {
+		return errors.New("remote source ref is undefined")
+	}
```

Comment:
the variable can't be `nil` at this point, or is it only a compliment check for the compiler?

applicable for the local source method